python/cudf/cudf/core/frame.py

```diff
@@ -1246,18 +1246,24 @@ def fillna(
                 for key, value in value.items()
             }
 
-        copy_data = self._data.copy(deep=True)
-
-        for name in copy_data.keys():
+        filled_data = {}
+        for col_name, col in self._data.items():
             should_fill = (
-                name in value
-                and not libcudf.scalar._is_null_host_scalar(value[name])
+                col_name in value
+                and col.can_fillna()
+                and not libcudf.scalar._is_null_host_scalar(value[col_name])
```

Comment:
Since `value` could be a `DataFrame` it would be good to save `value[col_name]` in a variable so that you can use it in computing `should_fill` and inside the `if` block to avoid invoking `__getitem__` twice.

Reply:
Done 👍